onyo/lib/pseudokeys.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from functools import partial
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from typing import (
+        Callable,
+        Dict,
+    )
+    from onyo.lib.items import Item
+
+
+@dataclass()
+class PseudoKey:
+    """"Defines a pseudo-key implementation."""
+    description: str
+    """description of what that pseudo-key is delivering"""
+    implementation: Callable
+    """Called by an `Item` when a pseudo-keys value is still undefined.
+
+    This Callable is expected to have a single parameter of type `Item`.
+    """
+
+
+def delegate(self: Item, attribute: str, *args, **kwargs):
+    # This is to avoid circular imports ATM.
+    return self.__getattribute__(attribute)(*args, **kwargs)
+
+
+PSEUDO_KEYS: Dict[str, PseudoKey] = {
+    'onyo.path.absolute': PseudoKey(description="Absolute path of the object.",
+                                    implementation=partial(delegate, attribute='get_path_absolute')
+                                    ),
+    'onyo.path.relative': PseudoKey(description="Path of the object relative to the repository root.",
+                                    implementation=partial(delegate, attribute='get_path_relative')
+                                    ),
+    'onyo.path.parent': PseudoKey(description="Path of the directory the object is in relative to the repository root.",
+                                  implementation=partial(delegate, attribute='get_path_parent')
+                                  ),
+    'onyo.path.file': PseudoKey(description="Path to the file containing an asset's YAML."
+                                            "Different from 'onyo.path.relative' in case of an asset directory.",
+                                implementation=partial(delegate, attribute='get_path_file')
+                                ),
+    'onyo.is.asset': PseudoKey(description="Whether the object is an asset.",
+                               implementation=partial(delegate, attribute='is_asset')
+                               ),
+    'onyo.is.directory': PseudoKey(description="Whether the object is a directory.",
+                                   implementation=partial(delegate, attribute='is_directory')
+                                   ),
+    'onyo.is.template': PseudoKey(description="Whether the object is a template.",
+                                  implementation=partial(delegate, attribute='is_template')
+                                  ),
+    'onyo.is.empty': PseudoKey(description="Whether the object is empty.",
+                               implementation=partial(delegate, attribute='is_empty')
+                               ),
+}
+r"""Pseudo-Keys are key names that are addressable but not in asset content.
+
+All ``PSEUDO_KEYS`` are reserved.
```

Comment:
This, in the near future will change. IMO, the entire `onyo.` namespace should be reserved (and nothing else).

Reply:
I guess, aliases are going to be a dynamic part of what is reserved, though. However, that may be a thing for fsck only.